sql/src/main/java/io/crate/planner/consumer/QueryThenFetchConsumer.java

```diff
@@ -138,43 +147,59 @@ public PlannedAnalyzedRelation visitQueriedTable(QueriedTable table, ConsumerCon
+            List<Projection> mergeProjections = new ArrayList<>();
             TopNProjection topNProjection = new TopNProjection(
                     MoreObjects.firstNonNull(table.querySpec().limit(), Constants.DEFAULT_SELECT_LIMIT),
                     table.querySpec().offset()
             );
-            List<Symbol> outputs = new ArrayList<>(2);
-            outputs.add(DEFAULT_DOC_ID_INPUT_COLUMN);
-            if (scoreSelected) {
-                outputs.add(DEFAULT_SCORE_INPUT_COLUMN);
-            }
-            topNProjection.outputs(outputs);
             mergeProjections.add(topNProjection);
 
-            // by default don't split fetch requests into pages/chunks,
-            // only if record set is higher than default limit
-            int bulkSize = FetchProjector.NO_BULK_REQUESTS;
-            if (topNProjection.limit() > Constants.DEFAULT_SELECT_LIMIT) {
-                bulkSize = Constants.DEFAULT_SELECT_LIMIT;
+            if (!outputsAreAllOrdered) {
+                List<Symbol> outputs = new ArrayList<>(collectSymbols.size());
+                for (int i = 0; i < collectSymbols.size(); i++) {
+                    outputs.add(new InputColumn(i, collectSymbols.get(i).valueType()));
+                }
+                topNProjection.outputs(outputs);
+                // by default don't split fetch requests into pages/chunks,
+                // only if record set is higher than default limit
+                int bulkSize = FetchProjector.NO_BULK_REQUESTS;
+                if (topNProjection.limit() > Constants.DEFAULT_SELECT_LIMIT) {
+                    bulkSize = Constants.DEFAULT_SELECT_LIMIT;
+                }
+                FetchProjection fetchProjection = new FetchProjection(
+                        DEFAULT_DOC_ID_INPUT_COLUMN, collectSymbols, outputSymbols,
+                        tableInfo.partitionedByColumns(),
+                        collectNode.executionNodes(),
+                        bulkSize,
+                        table.querySpec().isLimited());
+                mergeProjections.add(fetchProjection);
+            } else {
+                List<Symbol> outputs = new ArrayList<>(outputSymbols.size());
+                for (int i = 0; i < outputSymbols.size(); i++) {
+                    outputs.add(new InputColumn(i, outputSymbols.get(i).valueType()));
+                }
+                topNProjection.outputs(outputs);
             }
-            FetchProjection fetchProjection = new FetchProjection(
-                    DEFAULT_DOC_ID_INPUT_COLUMN, collectSymbols, outputSymbols,
-                    tableInfo.partitionedByColumns(),
-                    collectNode.executionNodes(),
-                    bulkSize,
-                    table.querySpec().isLimited());
-            mergeProjections.add(fetchProjection);
 
             MergeNode localMergeNode;
             if (orderBy != null) {
+                List<Symbol> inputs = new ArrayList<>(collectSymbols.size());
```

Comment:
please add a PlannerTest to ensure that the FetchProjection is not added in this case